services/cloudrun-service/pkg/cloudrun/cloudrun.go

```diff
@@ -0,0 +1,186 @@
+)
+
+const (
+	serviceLocationLabel       = "cloud.googleapis.com/location"
+	operationIdAnnotation      = "run.googleapis.com/operation-id"
+	serviceReady               = "Ready"
+	serviceConfigurationsReady = "ConfigurationsReady"
+	serviceRoutesReady         = "RoutesReady"
+)
+
+var (
+	runService *run.APIService
+)
+
+type ServiceReadyCondition struct {
+	Name     string
+	Revision int64
+	Status   string
+	Reason   string
+	Message  string
+}
+
+func (c ServiceReadyCondition) String() string {
+	return fmt.Sprintf("Service:%s, ObservedGeneration:%d, Ready:%s, Reason:%s, Message:%s", c.Name, c.Revision, c.Status, c.Reason, c.Message)
+}
+
+type serviceConfigError struct {
+	name             string
+	namespaceMissing bool
+	locationMissing  bool
+}
+
+func (p serviceConfigError) Error() string {
+	if p.namespaceMissing {
+		return fmt.Sprintf("Namespace is missing for service: %s", p.name)
+	}
+	if p.locationMissing {
+		return fmt.Sprintf("location is missing for service: %s", p.name)
+	}
+	return "Configuration error"
+}
+
+func Init(ctx context.Context) error {
+	var err error
+	runService, err = run.NewService(ctx)
+	return err
+}
+
+func Deploy(ctx context.Context, svc *run.Service) error {
+	serviceName := svc.Metadata.Name
+	// Get the full path of the project. Example: projects/<project-id>/locations/<region>
+	parent, err := getParent(svc)
+	servicePath := fmt.Sprintf("%s/services/%s", parent, serviceName)
+	if err != nil {
+		return err
+	}
+	req := runService.Projects.Locations.Services.List(parent)
+	resp, err := req.Do()
+	if err != nil {
+		return err
+	}
+	var serviceCallResp *run.Service
+	// If the service is already deployed before, then we need to call ReplaceService. Otherwise, we call Create.
+	if isPresent(resp.Items, serviceName) {
+		serviceCall := runService.Projects.Locations.Services.ReplaceService(servicePath, svc)
+		serviceCallResp, err = serviceCall.Do()
+		if err != nil {
+			return err
+		}
+	} else {
+		serviceCall := runService.Projects.Locations.Services.Create(parent, svc)
+		serviceCallResp, err = serviceCall.Do()
+		if err != nil {
+			return err
+		}
+	}
+	if err := waitForOperation(parent, serviceCallResp, 60); err != nil {
+		return err
+	}
+	getServiceCall := runService.Projects.Locations.Services.Get(servicePath)
+	serviceResp, err := getServiceCall.Do()
+	if err != nil {
+		return err
+	}
+	condition, err := GetServiceReadyCondition(serviceResp)
+	if err != nil {
+		return err
+	}
+	if condition.Status != "True" {
+		return fmt.Errorf("service not ready: %s", condition)
+	}
+	return nil
+}
+
+func GetServiceReadyCondition(serviceCallResponse *run.Service) (ServiceReadyCondition, error) {
+	//exhaustruct:ignore
+	serviceReadyCondition := ServiceReadyCondition{
+		Status:   "",
+		Name:     serviceCallResponse.Metadata.Name,
+		Revision: serviceCallResponse.Status.ObservedGeneration,
+	}
+	conditions := serviceCallResponse.Status.Conditions
+	for _, condition := range conditions {
```

Comment:
should we really check all conditions here, or should we just check the last one? Are they not sorted?

Reply:
I didn't want to count on the conditions being sorted. I think it is better to make sure that we are ready the correct condition.
Also there are a maximum of 6 conditions, so this does not have any impact on performance.